### PR TITLE
VPN Support

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Added support for Cisco ASA VPN and Fortinet VPN.

# Impacts
RADIUS Cli flow.

# Delete branch after merge
YES

# Checklist
- [no] Document the feature
- [no] Add unit tests
- [no] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* Cisco ASA VPN support (with the captive portal)
* Fortinet VPN support

### CODE REVIEW

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 1]
```diff
@@ -6347,6 +6347,93 @@ External portal SSID
 * Set the *Redirect Secret* to any passphrase of your choice
 * In the *RADIUS Configuration* section set the RADIUS server to point to your PacketFence server
 
+VPN Configuration
+-----------------
+
+Cisco ASA
+~~~~~~~~~
+
+Anyconnect
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: Anyconnect -> AnyConnect, everywhere.

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 2]
```diff
@@ -6347,6 +6347,93 @@ External portal SSID
...
+^^^^^^^^^^
+
+PacketFence support Cisco ASA VPN with anyconnect.
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: supports

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 3]
```diff
@@ -6347,6 +6347,93 @@ External portal SSID
...
+
+You can force the VPN user to authenticate first on the captive portal and based on the role of the device allow it and/or set set dynamic acl.
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: set set -> set, acl -> ACL

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 4]
```diff
@@ -6347,6 +6347,93 @@ External portal SSID
...
+
+In this example we assume that the Cisco ASA have 2 interfaces, one Management (192.168.2.1) where the VPN is activated and another one Registration (192.168.1.6) who is facing the PacketFence server (192.168.1.5).
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: have 2 -> has two
who is facing -> that is facing

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 5]
```diff
@@ -6347,6 +6347,93 @@ External portal SSID
...
+
+Before trying to configure PacketFence with the Cisco ASA first be sure that when you connect with Anyconnect and when the VPN is up that your device is able to reach internet.
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: internet -> Internet

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 6]
```diff
@@ -6347,6 +6347,93 @@ External portal SSID
...
+
+Cisco ASA
+~~~~~~~~~
+
+AnyConnect
+^^^^^^^^^^
+
+PacketFence supports Cisco ASA VPN with anyconnect.
+
+You can force the VPN user to authenticate first on the captive portal and based on the role of the device allow it and/or set dynamic ACL.
```
Comment (julsemaan) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: "force VPN users"

File: lib/pf/Switch.pm [hunk 7]
```diff
@@ -3066,6 +3074,34 @@ sub parseRequestUsername {
...
+
+sub parseVPNRequest {
+    my ( $self, $radius_request ) = @_;
+    my $logger = $self->logger;
+
+    my $client_ip       = ref($radius_request->{'Calling-Station-Id'}) eq 'ARRAY'
+                           ? clean_ip($radius_request->{'Calling-Station-Id'}[0])
+                           : clean_ip($radius_request->{'Calling-Station-Id'});
+
+    my $user_name       = $radius_request->{'PacketFence-UserNameAttribute'} || $radius_request->{'TLS-Client-Cert-Subject-Alt-Name-Upn'} || $radius_request->{'TLS-Client-Cert-Common-Name'} || $radius_request->{'User-Name'};
```
Comment (julsemaan) [inline lib/pf/Switch.pm]: use the new username parsing subroutine to extract the username

File: lib/pf/Switch/Cisco/ASA.pm [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    my ( $self, $mac, $is_dot1x ) = @_;
+    my $logger = $self->logger;
+
+    if ( !$self->isProductionMode() ) {
+        $logger->info("not in production mode... we won't perform deauthentication");
+        return 1;
+    }
+
+    $logger->debug("Change the state of $mac using RADIUS CoA");
+    # TODO push Login-User => 1 (RFC2865) in pf::radius::constants if someone ever reads this
```
Comment (julsemaan) [inline lib/pf/Switch/Cisco/ASA.pm]: Is that comment really accurate ?

File: lib/pf/Switch/Cisco/ASA.pm [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+sub parseVPNRequest {
+    my ( $self, $radius_request ) = @_;
+    my $logger = $self->logger;
+
+    my $client_mac;
+    my $client_ip       = ref($radius_request->{'Calling-Station-Id'}) eq 'ARRAY'
+                           ? clean_ip($radius_request->{'Calling-Station-Id'}[0])
+                           : clean_ip($radius_request->{'Calling-Station-Id'});
+
+    my $user_name       = $radius_request->{'PacketFence-UserNameAttribute'} || $radius_request->{'TLS-Client-Cert-Subject-Alt-Name-Upn'} || $radius_request->{'TLS-Client-Cert-Common-Name'} || $radius_request->{'User-Name'};
```
Comment (julsemaan) [inline lib/pf/Switch/Cisco/ASA.pm]: use the new parse username sub

File: lib/pf/Switch/Fortinet/FortiGate.pm [hunk 10]
```diff
@@ -191,6 +192,77 @@ sub getVersion {
...
+
+sub parseVPNRequest {
+    my ( $self, $radius_request ) = @_;
+    my $logger = $self->logger;
+
+    my $client_ip       = ref($radius_request->{'Calling-Station-Id'}) eq 'ARRAY'
+                           ? clean_ip($radius_request->{'Calling-Station-Id'}[0])
+                           : clean_ip($radius_request->{'Calling-Station-Id'});
+
+    my $user_name       = $radius_request->{'PacketFence-UserNameAttribute'} || $radius_request->{'TLS-Client-Cert-Subject-Alt-Name-Upn'} || $radius_request->{'TLS-Client-Cert-Common-Name'} || $radius_request->{'User-Name'};
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet/FortiGate.pm]: use the new parse username

File: lib/pf/radius/rest.pm [hunk 11]
```diff
@@ -71,7 +71,7 @@ sub format_request {
     my ($request) = @_;
     # transform the request according to what radius_authorize expects
     my %remapped_radius_request = map {
-        $_ => ( exists $ALLOWED_MULTIPLE{$_} ) ? $request->{$_}{value} : $request->{$_}{value}[0]
```
Comment (julsemaan) [inline lib/pf/radius/rest.pm:74]: Should we just remove the ALLOWED_MULTIPLE definition then or is it used elsewhere?

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 12]
```diff
@@ -6347,6 +6347,93 @@ External portal SSID
...
+VPN Configuration
+-----------------
+
+Cisco ASA
+~~~~~~~~~
+
+AnyConnect
+^^^^^^^^^^
+
+PacketFence supports Cisco ASA VPN with anyconnect.
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: anyconnect -> AnyConnect

**Comment by extrafu** [comment]
@julsemaan Ready to be merged?

**Comment by extrafu** [comment]
2wowjx

**Comment by julsemaan** [comment]
https://www.youtube.com/watch?v=LBduNcf1eQc

**Pull request merged**